Old school version (#64)
oda-hub/hugo-odahub: content/docs/guide-development.md

# [MMODA](https://www.astro.unige.ch/mmoda) workflow development Guide

> The [MMODA](https://www.astro.unige.ch/mmoda) platform provides access to the Astronomical Open Research Data Analysis Services (ORDAS). Good fraction of these services follow a simple scheme: they  

Author: hm, I would change somewhat but removing "they" breaks the sentence which continues into the list.

> The [MMODA](https://www.astro.unige.ch/mmoda) platform provides access to the Astronomical Open Research Data Analysis Services (ORDAS). Good fraction of these services follow a simple scheme: they  

Author: ```suggestion
The [MMODA](https://www.astro.unige.ch/mmoda) platform provides access to the Astronomical Open Research Data Analysis Services (ORDAS). Good fraction of these services follow a simple scheme, they:  
```

* access publicly available external astronomical data archives to fetch data relevant to specific source or source catalog,

> * transform the fetched data using a workflow based on a Python notebook, to drive a data product and

Author: ```suggestion
* transform the fetched data using a workflow based on a Python notebook, to derive a data product and
```

> The users of [MMODA](https://www.astro.unige.ch/mmoda) are encouraged to become its developers and add new ORDAS. This help page provides a step-by step instructions on how to add new services to MMODA.

Reviewer: Duplicating line 9

Author: Could you put a suggestion?

* display a preview of the data product on the [MMODA](https://www.astro.unige.ch/mmoda) frontend and/or return the data product to the user via  Python [API](https://github.com/oda-hub/oda_api)

> The users of [MMODA](https://www.astro.unige.ch/mmoda) are encouraged to become its developers and add new ORDAS. This help page provides a step-by step instructions on how to add new services to [MMODA](https://www.astro.unige.ch/mmoda) .

Author: it means to add (introduce, construct) new (previously not available) ORDAS (Open Research Data Analysis Services, see it spelled out above).

It seems clear to me but if not, please re-open and propose another formulation.

> The users of [MMODA](https://www.astro.unige.ch/mmoda) are encouraged to become its developers and add new ORDAS. This help page provides a step-by step instructions on how to add new services to [MMODA](https://www.astro.unige.ch/mmoda) .

Reviewer 2: I would simply add: 'add a new ORDAS service'

> The users of [MMODA](https://www.astro.unige.ch/mmoda) are encouraged to become its developers and add new ORDAS. This help page provides a step-by step instructions on how to add new services to [MMODA](https://www.astro.unige.ch/mmoda) .

Author: "S" in ORDAS is "service". So "add a new ORDAS service" == "add a new Open Research Data Analysis Service service" - not nice!

> Workflows are all things that can be computed, broadly speaking. For reproducibility, we want our workflows to be repeatable: producing the same output every time they are computed. This looks easy enough in first approximation, but might be harder to achieve than it seems when the workflow relies on external data and compute resources  and has to yield an "up-and-running" ORDAS in an ever-evolving compute environment. This help page is also aimed at helping the developers in ensuring reproducibility and reusability of their workflows converted to ORDAS.

Author: this misses a bit the point which was there, about that to reproduce it is not enough to just repeat, to properly reproduce (and hence reuse with trust) it is almost always necessary to change something.

### How to designate input parameters and output cells of the notebook 

In [MMODA](https://www.astro.unige.ch/mmoda) we use the approach of [papermill](https://papermill.readthedocs.io/en/latest/usage-parameterize.html#designate-parameters-for-a-cell)) to **tag** the notebook cells that contain the input parameters and the outputs. In your notebook, you may create two dedicated cells with the "parameters" and "outputs" tags. In the Jupyter Lab environment this can be done by clicking on the cogwheel sign on the right top, red arrow in the image below, and adding new tag as pointed by the second red arrow:

> ![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp1.png)

Author: ?

> ![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp1.png)

Reviewer 2: I just changed the link and I think github is trying to load the image
this was the intended:

`![](tmp1.png)`

### How to define input parameters in the dedicated parameters cell 

The variables defined in the dedicated "parameters" cell, will be the input parameters of the workflow. They will be visualized in the frontend of the service and it will be possible to provide these parameters via the service [API](https://github.com/oda-hub/oda_api). For example, in the image of the parameters cell in our example (see above)
  * the names of the declared variables will be used as parameter names in the [MMODA](https://www.astro.unige.ch/mmoda) service (except the **default** parameters, see below). For example, the `Source_region_radius` variable will be visible in the frontend as a query parameter with the same name. It will appear with a default value assigned to it (`2.` in the example notebook). 
  * if not annotated, the types of the inputs parameters are determined based on the parameter default value (would be `float` for the `Source_region_radius` parameter).

>   * otherwise, it is possible to customize the parameter by adding annotation the input parameter with an MMODA [ontology](https://odahub.io/docs/guide-ontology) item as a comment (after the hash sign, `#http://odahub.io/ontology#AngleDegrees` in the reference example of the `Source_region_radius` parameter. This may be useful for checking the validity of the user inputs. For example, the sky angle in degrees (defined by the `#http://odahub.io/ontology#AngleDegrees`) should be a float number and can take values between 0 and 360. 

Author: I think it's basically the same, and if Andrii find the first, longer, version better - let's keep it since he better represents.

the same result every time but still be reproducible (see motivation on [the difference between reproducibility and repeatability](https://github.com/volodymyrss/reproducibility-motivation/)). 

 ## 2) Make the notebook available for deployment on [MMODA](https://www.astro.unige.ch/mmoda/) via [renkulab.io](https://renkulab.io/)

The parameterized workflow formulated as a Python notebook can be converted into a service provided by [MMODA](https://www.astro.unige.ch/mmoda/)  by a bot that scans a specific location `astronomy/mmoda` in the project directory on the  [renkulab.io](https://renkulab.io/) collaborative platform. Creating a new project in this directory will make it visible for the bot. In our example of Fermi/LAT lightcurve workflow, it is in the [fermi](https://renkulab.io/projects/astronomy/mmoda/fermi) subdirectory of `astronomy/mmoda`.

To proceed along this way, you first you need to make sure your notebook runs correctly in the  [renkulab.io](https://renkulab.io/) environment. You can start a new project in the `astronomy/mmoda` by clicking on the "Creat new project" button.

![image](tmp4.png)

> You will need to choose the name of the new project. This name will be the name of your service appearing at the [MMODA](https://www.astro.unige.ch/mmoda) frontend and discoverable via [MMODA](https://www.astro.unige.ch/mmoda) API. Place your project in the `astronomy/mmoda` namespace by specifying this in the "Namespace" field as shown below:

Reviewer 2: ```suggestion
You will need to choose the name of the new project. This name will be the name of your service appearing at the [MMODA](https://www.astro.unige.ch/mmoda) frontend and discoverable via [MMODA](https://www.astro.unige.ch/mmoda) API. Place your project in the `astronomy/mmoda` namespace by specifying this in the "Namespace" field as shown below. Alternatively, a new project can be created using the following [URL](https://renkulab.io/projects/new?data=eyJ1cmwiOiJodHRwczovL2dpdGh1Yi5jb20vb2RhLWh1Yi9jb250cmlidXRlZC1wcm9qZWN0LXRlbXBsYXRlcyIsInJlZiI6Im1haW4iLCJ0ZW1wbGF0ZSI6IkN1c3RvbS9tbW9kYSJ9) which will set a predefined, MModa specific, session template, can be created after having filled in some basic information.
```

Author: Looks like this does not actually set a namespace.

Reviewer 2: I am struggling in trying to set the namespace, I get a warning: `The namespace "astronomy/mmoda" is not available.` 

Any ideas? Could it be permission related?

Author: We had a problem with setting namespace discussed with @rokroskar before. I am not sure what's the status of this option now. If it does not work, let's avoid the link (I guess that's why I removed it eventually and left manual instructions).

Reviewer 2: ok, I also tried to play around with the variables of the custom template, without succeeding.
I saw that I was not allowed to set any namespace different from my own username

no link then...

![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp5.png)

> Choose the "Python 3" template for the project and then click "Create project" button:

Author: At the moment MMODA templates are not available in the most easy way. Once they are, we can suggest them https://github.com/oda-hub/hugo-odahub/issues/65

> Choose the "Python 3" template for the project and then click "Create project" button:

Author: I was sure I had a link like that in here but apparently not. Anyway, we probably need both, a link to create with values substituted as well as instruction which Andrii put. Please put a suggestion.

![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp6.png)

To start working on the newly created project, you can launch an interactive Jupyte lab session by clicking on the "Start" button:

![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp7.png)

Once in the Jupyter lab environment, you can update the project by uploading the notebook that you intend to promote to a service:

![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp8.png)

> Your notebook most probably imports some python packages that are not installed by default in a generic Python environment on [renkulab.io](https://renkulab.io). To add necessary packages, you need to update the `requirements.txt` file with the packages you need:

Reviewer: Also, environment.yml

> Your notebook most probably imports some python packages that are not installed by default in a generic Python environment on [renkulab.io](https://renkulab.io). To add necessary packages, you need to update the `requirements.txt` file with the packages you need:

Author: ```suggestion
Your notebook most probably imports some python packages that are not installed by default in a generic Python environment on [renkulab.io](https://renkulab.io). To add necessary packages, you need to update the `requirements.txt` and possibly `environment.yml` files with the packages you need:
```

![image](https://github.com/oda-hub/hugo-odahub/blob/master/content/docs/tmp15_1.png)

Note that some of the input parameters in the example of the Fermi/LAT [Lightcurve.ipynb](https://renkulab.io/projects/astronomy/mmoda/fermi/files/blob/Lightcurve.ipynb) notebook appear as multiple choice parameters with pre-defined values, while others are query fields. For some of the parameters, units are specified just below the query window. The names of the parameters are the names of the variables defined in the `parameters` cell of the notebook (see the screenshot of the parameters cell above in section (1) of this help page. 

If the `outputs` cell of your notebook contains multiple data products, they will be shown as a list at the [MMODA](https://www.astro.unige.ch/mmoda) frontend, as shown above. The names of the list items correspond to the names of the variables defined in the `outputs` cell. Each item of the list can be previewed or downloaded by clicking on the "View" button. The preview will depend on the type of the data product that has been specified after the comment  hash `#` tag in the `outputs` cell. 

> You can look at different examples of the notebooks converted to services in the `astronomy/mmoda` domain on renkulab.io, to see how to format the inputs and outputs. You can also experiment with further possibilities exploring the [ontology](https://odahub.io/ontology/) of the [MMODA](https://www.astro.unige.ch/mmoda) parameters and data products. 

Reviewer: https://renkulab.io/projects/astronomy/mmoda/mmoda-nb2workflow-example could be mentioned as a repo containing examples for all (or most of) possible output types

Author: ```suggestion
You can explore different examples of the notebooks converted to services in the `astronomy/mmoda` domain on renkulab.io, to see how to format the inputs and outputs. If unsure, first take a look on [this simple repository](https://renkulab.io/projects/astronomy/mmoda/mmoda-nb2workflow-example). You can also experiment with further possibilities exploring the [ontology](https://odahub.io/ontology/) of the [MMODA](https://www.astro.unige.ch/mmoda) parameters and data products. 
```
